src/main/java/io/nats/client/impl/NatsMessage.java

```diff
@@ -26,103 +26,75 @@
             } else {
-                idx--;
-                lengthBytes[idx] = digits[0];
-            }
-
-            // Build the array
-            int len = 4 + subject.length() + 1 + (lengthBytes.length - idx);
-
-            if (replyTo != null) {
                 len += replyTo.length() + 1;
             }
+        }
+        if (utf8mode) {
+            len += fastUtf8Length(subject) + 1;
+            charset = StandardCharsets.UTF_8;
+        } else {
+            len += subject.length() + 1;
+            charset = StandardCharsets.US_ASCII;
+        }
+        this.protocolBytes = ByteBuffer.allocate(len);
+        protocolBytes.put((byte)'P').put((byte)'U').put((byte)'B').put((byte)' ');
+        protocolBytes.put(subject.getBytes(charset));
+        protocolBytes.put((byte)' ');
 
-            this.protocolBytes = new byte[len];
-
-            // Copy everything
-            int pos = 0;
-            protocolBytes[0] = 'P';
-            protocolBytes[1] = 'U';
-            protocolBytes[2] = 'B';
-            protocolBytes[3] = ' ';
-            pos = 4;
-            pos = copy(protocolBytes, pos, subject);
-            protocolBytes[pos] = ' ';
-            pos++;
-
-            if (replyTo != null) {
-                pos = copy(protocolBytes, pos, replyTo);
-                protocolBytes[pos] = ' ';
-                pos++;
-            }
+        if (replyTo != null) {
+            protocolBytes.put(replyTo.getBytes(charset));
+            protocolBytes.put((byte)' ');
+        }
 
-            System.arraycopy(lengthBytes, idx, protocolBytes, pos, lengthBytes.length - idx);
+        if (size > 0) {
+            int base = protocolBytes.limit();
+            for (int i = size; i > 0; i /= 10) {
+                base--;
+                protocolBytes.put(base, (byte)(i % 10 + 0x30));
+            }
+        } else {
+            protocolBytes.put((byte)0x30);
         }
+        protocolBytes.clear();
+    }
 
-        this.sizeInBytes = this.protocolBytes.length + data.length + 4;// for 2x \r\n
+    NatsMessage(String subject, String replyTo, byte[] data, boolean utf8mode) {
+        this(subject, replyTo, ByteBuffer.wrap(data), utf8mode);
     }
 
     // Create a protocol only message to publish
     NatsMessage(CharBuffer protocol) {
-        ByteBuffer byteBuffer = StandardCharsets.UTF_8.encode(protocol);
-        this.protocolBytes = Arrays.copyOfRange(byteBuffer.array(), byteBuffer.position(), byteBuffer.limit());
-        Arrays.fill(byteBuffer.array(), (byte) 0); // clear sensitive data
-        this.sizeInBytes = this.protocolBytes.length + 2;// for \r\n
+        if (protocol.remaining() == 0) {
+            this.protocolBytes = ByteBuffer.allocate(0);
+        } else {
+            protocol.mark();
+            this.protocolBytes = ByteBuffer.allocate(fastUtf8Length(protocol));
+            protocol.reset();
+            StandardCharsets.UTF_8.newEncoder().encode(protocol, this.protocolBytes, true);

```

Comment:
Could you use a static instance rather than allocating every new msg?

Reply:
I recall I tested that and this was a good bit faster in benchmarks, not 100% sure why, best guess is having multiple threads accessing the same static encoder instance causes cache flushing or something. I plan to refactor this anyways down the line once I refactor some prerequisite code.

Reply:
Yep, verified this has a major performance impact.
Static instance:
```
5:59:11 PM: Executing task 'MessageProtocolCreationBenchmark.main()'...

> Task :compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava UP-TO-DATE
> Task :processTestResources UP-TO-DATE
> Task :testClasses UP-TO-DATE

> Task :MessageProtocolCreationBenchmark.main()
### Running benchmarks with 50,000,000 messages.

### Total time to create 50,000,000 non-utf8 messages for sending was 1,835 ms
	36.718829 ns/op
	27,233,984.034 op/sec

### Total time to create 50,000,000 utf8 messages for sending was 3,552 ms
	71.053063 ns/op
	14,073,988.718 op/sec

### Total time to create 50,000,000 a protocol message was 341 ms
	6.833834 ns/op
	146,330,744.581 op/sec

BUILD SUCCESSFUL in 6s
4 actionable tasks: 1 executed, 3 up-to-date
5:59:17 PM: Task execution finished 'MessageProtocolCreationBenchmark.main()'.
```
non-static(current) implementation:
```
5:58:33 PM: Executing task 'MessageProtocolCreationBenchmark.main()'...


> Task :compileJava

> Task :processResources NO-SOURCE
> Task :classes
> Task :compileTestJava UP-TO-DATE
> Task :processTestResources UP-TO-DATE
> Task :testClasses UP-TO-DATE
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

> Task :MessageProtocolCreationBenchmark.main()
### Running benchmarks with 50,000,000 messages.

### Total time to create 50,000,000 non-utf8 messages for sending was 1,802 ms
	36.054858 ns/op
	27,735,513.355 op/sec

### Total time to create 50,000,000 utf8 messages for sending was 2,841 ms
	56.833210 ns/op
	17,595,346.163 op/sec

### Total time to create 50,000,000 a protocol message was 334 ms
	6.693376 ns/op
	149,401,436.293 op/sec

BUILD SUCCESSFUL in 5s
4 actionable tasks: 2 executed, 2 up-to-date
5:58:39 PM: Task execution finished 'MessageProtocolCreationBenchmark.main()'.
```

Reply:
Wow, would have lost a bet on that one.  Thanks!

Reply:
Also noticed this [here](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CharsetEncoder.html):
```
Instances of this class are not safe for use by multiple concurrent threads.
```